services/agents/channel_test.go

```diff
@@ -76,188 +80,277 @@ func setup(t *testing.T, connect func(agent.Agent_ConnectServer) error, expected
 	require.NoError(t, err, "failed to dial server")
 	stream, err := agent.NewAgentClient(cc).Connect(ctx)
 	require.NoError(t, err, "failed to create stream")
-	channel := NewChannel(stream)
 
 	teardown := func(t *testing.T) {
+		require.NotNil(t, channel, "Test exited before first message reached connect handler.")
+
 		err := channel.Wait()
 		assert.Contains(t, expected, errors.Cause(err), "%+v", err)
 
 		server.GracefulStop()
 		cancel()
 	}
 
-	return channel, cc, teardown
+	return stream, cc, teardown
 }
 
 func TestAgentRequest(t *testing.T) {
 	const count = 50
 	require.True(t, count > serverRequestsCap)
 
-	connect := func(stream agent.Agent_ConnectServer) error {
+	var channel *Channel
+	connect := func(ch *Channel) error { //nolint:unparam
+		channel = ch // store to check metrics below
+
 		for i := uint32(1); i <= count; i++ {
-			msg, err := stream.Recv()
-			require.NoError(t, err)
+			msg := <-ch.Requests()
+			require.NotNil(t, msg)
 			assert.Equal(t, i, msg.Id)
-			require.NotNil(t, msg.GetQanData())
+			assert.NotNil(t, msg.GetQanData())
 
-			err = stream.Send(&agent.ServerMessage{
+			ch.SendResponse(&agent.ServerMessage{
 				Id: i,
 				Payload: &agent.ServerMessage_QanData{
 					QanData: new(agent.QANDataResponse),
 				},
 			})
-			assert.NoError(t, err)
 		}
 
+		assert.Nil(t, <-ch.Requests())
 		return nil
 	}
 
-	channel, _, teardown := setup(t, connect, io.EOF)
+	stream, _, teardown := setup(t, connect, io.EOF) // EOF = server exits from handler
 	defer teardown(t)
 
 	for i := uint32(1); i <= count; i++ {
-		resp := channel.SendRequest(&agent.AgentMessage_QanData{
-			QanData: new(agent.QANDataRequest),
+		err := stream.Send(&agent.AgentMessage{
+			Id: i,
+			Payload: &agent.AgentMessage_QanData{
+				QanData: new(agent.QANDataRequest),
+			},
 		})
-		assert.NotNil(t, resp)
+		require.NoError(t, err)
+
+		msg, err := stream.Recv()
+		require.NoError(t, err)
+		assert.Equal(t, i, msg.Id)
+		assert.NotNil(t, msg.GetQanData())
 	}
+
+	err := stream.CloseSend()
+	assert.NoError(t, err)
+
+	// check metrics
```

Comment:
maybe it's better to check metrics in separate test?

Reply:
Maybe. I don't know yet. Probably when we will have more metrics.